src/main/java/com/aws/greengrass/authorization/AuthorizationModule.java

```diff
@@ -42,17 +47,61 @@ public void addPermission(String destination, Permission permission) throws Auth
                 || Utils.isEmpty(permission.getOperation())) {
             throw new AuthorizationException("Invalid arguments");
         }
-        // resource as null is ok, but it should not be empty
         String resource = permission.getResource();
-        if (resource != null && Utils.isEmpty(resource)) {
-            throw new AuthorizationException("Resource cannot be empty");
-        }
+        String validatedResource = validateResource(resource);
         resourceAuthZCompleteMap.get(destination).get(permission.getPrincipal()).get(permission.getOperation()).add(
-                permission.getResource());
+                resource);
```

Comment:
How is this map used? Why it doesn't need the validated resource?

Reply:
This map maintains the combination of Destination, principal, operation and the resource trie with it.
ValidateResource also parses and removes escaping to give out raw resource. But the add method for the trie needs the original resource as it needs that info to differentiate real wildcard and escaped wildcard chars

Reply:
Thanks for explaining.